File: Tests/ModTrial/mote32.pas
unit mote32;
{  MOTE32 - A small-state CSPRNG and Stream Cipher
   MOTE32 is a MOTE with a 32+4-word internal state
   MOTE32 may be seeded with a 1024-bit key
   MOTE32 Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
   MOTE32 is inspired by Bob Jenkins' PRNG insights (Public Domain). }
{$mode delphi}
{$inline on}
//are we testing?
{ $define TEST}
//verbose test output
{ $define VERBOSE}

interface

// initialize/reset MOTE32
procedure moReset;
// obtain a MOTE32 pseudo-random value in [0..2**32]
function moRandom: Cardinal;
// seed MOTE32 with a 1024-bit block of 4-byte words (Bob Jenkins method) 
procedure moSeedW(seed: string; rounds: integer);
// MOTE32 # of bits internal state
function moStateBits: Cardinal;
// MOTE32 expected cycle length
function moCycle: Cardinal;
// MOTE32 Name
function moName: string;

implementation
{$ifdef TEST}uses StrUtils;{$endif}

const 	NAME = 'MOTE32';
		STSZ = 32;
		STM1 = STSZ-1;
		STBYTES = STSZ*4;
		STBITS = 128+STBYTES*8;
		// 2**32/phi, where phi is the golden ratio
		GOLDEN   = $9e3779b9;
		FLEASEED = $f1ea5eed;

		MODU = 26;
		START= 65;
		
// MOTE32 STATE
var		rsl: 	array[0..STSZ] of Cardinal;
		state: 	array[0..STSZ] of Cardinal;
		b,c,d,e,rcnt: Cardinal;

// MOTE32 ROT switcher
var			ri: Cardinal = 0;
type 		TRsw = record iii: Cardinal; jjj: Cardinal; kkk: Cardinal end;
type 		ARsw = array[0..3] of TRsw;
const		rsw: ARsw = ( 
(iii:30;  jjj: 4; kkk:20;), // avalanche: 19.00 bits
(iii:11;  jjj: 2; kkk:26;), // avalanche: 18.75 bits
(iii:19;  jjj:19; kkk:14;), // avalanche: 18.75 bits
(iii: 6;  jjj:16; kkk:12;)  // avalanche: 18.25 bits
);

	
function rot(var x: Cardinal; const k: Cardinal): Cardinal; inline;
	begin
		rot := (x shl k) or (x shr (32-k));
	end;

	
{$ifdef TEST}
var bcnt: Cardinal = 0;
procedure statepeek; 
	var i: Cardinal;
	begin
		inc(bcnt);
		Writeln(bcnt:3,') mote32 using rsw[',ri,']...');
		for i:=0 to STM1 do begin
			{$ifdef VERBOSE}
			Writeln('rsl ',
				i:3,': ',rsl[i]:11,chr(rsl[i] mod MODU + START):2,
					dec2numb((rsl[i] and 255),2,16):3,'  | state ',
				i:3,': ',state[i]:11,chr(state[i] mod MODU + START):2,
					dec2numb((state[i] and 255),2,16):3);
			{$endif}
		end;
		Writeln('     b = ',b:11,chr(b mod MODU+START):2,dec2numb((b and 255),2,16):3);     
		Writeln('     c = ',c:11,chr(c mod MODU+START):2,dec2numb((c and 255),2,16):3);    
		Writeln('     d = ',d:11,chr(d mod MODU+START):2,dec2numb((d and 255),2,16):3);
		Writeln('     e = ',e:11,chr(e mod MODU+START):2,dec2numb((e and 255),2,16):3);
	end;
{$endif}


// MOTE32 is filled every 32 rounds
procedure mote32;
    var i: Cardinal;
	begin
		for i:=0 to STM1 do begin
			state[c and STM1] := e;
			b := c xor ((e shr rsw[ri].iii) or (d shl rsw[ri].jjj));
			c := d - rot(b,rsw[ri].kkk);
			d := state[i] + b;
			e := c + d;
			rsl[i] := c;
		end;
		{$ifdef TEST}
		statepeek;
		{$endif}
		ri := (d and 3);
	end;
	
	
// initialize/reset MOTE32 (obligatory)
procedure moReset;
	var i: Cardinal;
	begin
		rcnt:=0;
		b := FLEASEED; c := b; d:=c; e:=d;
		for i:=0 to STM1 do begin
			state[i] := GOLDEN; rsl[i] := 0;
		end;
	end;


// obtain a MOTE pseudo-random value in [0..2**32]
function moRandom: Cardinal;
	begin
		moRandom := rsl[rcnt];
		inc(rcnt);
		if (rcnt=STSZ) then begin
			mote32;
			rcnt := 0;
		end;
	end;


// mix in all the key and state bytes
procedure mix;
	var i: Cardinal;
	begin
		for i:=0 to STM1 do begin
			b := c xor ((e shr rsw[ri].iii) or (d shl rsw[ri].jjj));
			c := d - rot(b,rsw[ri].kkk);
			d := state[i] + b;
			e := c + d;
			state[i] := e;
		end;
	end;

	
// seed MOTE32 with a 1024-bit block of 4-byte words (Bob Jenkins method) 
procedure moSeedW(seed: string; rounds: integer);
	var i,l: Cardinal;
		p: pointer;
	begin
		p:=@state[0];
		l:=Length(seed);
		if (l>STBYTES) then l:=STBYTES;
		moReset;
		for i:=0 to l-1 do
			byte((p+i)^) := byte(seed[i+1]);
		mix;
		mote32;
		for i:=1 to rounds do moRandom;  
	end;


// MOTE32 # of bits internal state
function moStateBits: Cardinal;
	begin
		moStateBits := STBITS;
	end;


// MOTE32 expected cycle length
function moCycle: Cardinal;
	begin
		moCycle := (STBITS+1) div 2;
	end;


// MOTE32 Name
function moName: string;
	begin
		moName := NAME;
	end;

end.

File: C/MOTET/mote32.c
/* MOTE32 - A small-state CSPRNG and Stream Cipher
   MOTE32 is a MOTE with a 32+4-word internal state
   MOTE32 may be seeded with a 1048-bit key
   MOTE32 Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
   MOTE32 is inspired by Bob Jenkins' PRNG insights (Public Domain).
*/
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include "mote32.h"

// MOTE defines
//#define TEST
//#define VERBOSE
#define NAME "MOTE32"
#define STSZ 32
#define STM1 STSZ-1
#define STBYTES STSZ*4
#define STBITS 128+STBYTES*8

/* 2**32/phi, where phi is the golden ratio */
#define GOLDEN   0x9e3779b9
#define FLEASEED 0xf1ea5eed
#define rot(x,k) ((x<<k)|(x>>(32-k)))

static u4 rcnt=0;
static u4 state[STSZ], rsl[STSZ], b, c, d, e;

// MOTE32 ROT switcher
typedef struct Rsw { u4 iii; u4 jjj; u4 kkk; };
static u4 ri=0;
static const struct Rsw rsw[4] = {
30,  4, 20, // avalanche: 19.00 bits
11,  2, 26, // avalanche: 18.75 bits
19, 19, 14, // avalanche: 18.75 bits
 6, 16, 12  // avalanche: 18.25 bits
};

#ifdef TEST
static void statepeek(void);
#endif


// MOTE32 is refilled every 32 rounds
static void mote32(void) {
	register u4 i;
	for (i=0; i<STSZ; i++) {
		state[c & STM1] = e;
		b = c ^ (e >> rsw[ri].iii | d << rsw[ri].jjj);
		c = d - rot(b,rsw[ri].kkk);
		d = state[i] + b;
		e = c + d;
		rsl[i] = c;
	}
	#ifdef TEST
	statepeek();
	#endif
	ri = d & 3;
}


// obtain a MOTE pseudo-random value in [0..2**32]
u4 mote32_Random(void) {
	u4 r = rsl[rcnt];
	++rcnt;
	if (rcnt==STSZ) {
		mote32();
		rcnt = 0;
	}
	return r;
}


void mote32_Reset(void) {
	register u4 i,r;
	rcnt = 0; ri = 0;
	b = c = d = e = FLEASEED;
	for (i=0; i<STSZ; i++) { state[i]=GOLDEN; rsl[i]=0; }
}


// mix in all the key and state bytes
static void mix(void) {
	register u4 i;
	for (i=0; i<STSZ; i++) {
		b = c ^ (e >> rsw[ri].iii | d << rsw[ri].jjj);
		c = d - rot(b,rsw[ri].kkk);
		d = state[i] + b;
		e = c + d;
		state[i] = e;
	}
}


// seed MOTE with a 1024-bit block of 4-byte words (Bob Jenkins method) 
void mote32_SeedW(char *seed, int rounds)
{
	register u4 i,l;
	char s[STBYTES*2];
	l=strlen(seed);
	if (l>STBYTES) l=STBYTES;
	memset(s,0,l+1);
	strcpy(s,seed);
	mote32_Reset();
	memcpy((char *)state, (char *)s, l);
	mix();
	mote32();
	for (i=0; i<rounds; i++) mote32_Random();  
}


// MOTE32 # of bits internal state
u4 mote32_StateBits(void) {
	return STBITS;
}


// MOTE32 expected cycle length
u4 mote32_Cycle(void) {
	return (STBITS+1)/2;
}


// MOTE32 maximum key length (bits)
u4 mote32_KeyLength(void) {
	return STBYTES*8;
}


// MOTE32 Name
char* mote32_Name(void) {
	return NAME;
}


#ifdef TEST
static u4 bcnt=0;
static void statepeek(void) {
	register u4 i;
	++bcnt;
	printf("%3u) mote32 using rsw[%1u]...\n",bcnt,ri);
	for (i=0; i<STSZ; i++) {
		#ifdef VERBOSE
		printf("rsl %3u: %11u %c %02X  | state %3u: %11u %c %02X\n",
			i,rsl[i],rsl[i] % 26 + 'A',rsl[i] & 255,
			i,state[i],state[i] % 26 + 'A',state[i] & 255);
		#endif
	}
	printf("     b = %11u %c %02X\n     c = %11u %c %02X\n     d = %11u %c %02X\n     e = %11u %c %02X\n",
		b,b % 26+'A',b & 255,c,c % 26+'A',c & 255,d,d % 26+'A',d & 255,e,e % 26+'A',e & 255);
}
#endif
